csc2/macc_so.c

```diff
@@ -3044,142 +3040,118 @@ int dyns_get_table_field_option(char *tag, int fidx, int option,
-                if (*value_type == CLIENT_BYTEARRAY && vbsz >= length) {
-                    memcpy(valuebuf, bytes, length);
-                    *value_sz = length;
-                    return 0;
-                }
+            return -1;
+        }
+        case CLIENT_BYTEARRAY: {
+            int *bytes;
+            int length;
+            bytes = (int *)f->value.byteval;
+            if (!bytes) {
+                csc2_error("dyns_get_table_field_option: null byteval\n");
                 return -1;
             }
-            case CLIENT_CSTR: {
-                if ((*value_type == CLIENT_CSTR || *value_type == CLIENT_PSTR ||
-                     *value_type == CLIENT_VUTF8) &&
-                    vbsz >
-                        strlen(tables[tidx].sym[fidx].fopts[i].value.strval)) {
-                    bzero(valuebuf,
-                          strlen(tables[tidx].sym[fidx].fopts[i].value.strval) +
-                              1);
-                    memcpy(
-                        valuebuf, tables[tidx].sym[fidx].fopts[i].value.strval,
-                        strlen(tables[tidx].sym[fidx].fopts[i].value.strval));
-                    *value_sz =
-                        strlen(tables[tidx].sym[fidx].fopts[i].value.strval);
-                    return 0;
-                } else if (*value_type == CLIENT_DATETIME ||
-                           *value_type == CLIENT_DATETIMEUS) {
-                    memcpy(
-                        valuebuf, tables[tidx].sym[fidx].fopts[i].value.strval,
-                        strlen(tables[tidx].sym[fidx].fopts[i].value.strval));
-                    *value_sz =
-                        strlen(tables[tidx].sym[fidx].fopts[i].value.strval);
-                    return 0;
-                } else if (*value_type == CLIENT_BYTEARRAY &&
-                           vbsz >= tables[tidx].sym[fidx].szof) {
-                    /* There are several production databases that try to
-                     * specify a default load/store for a byte array using a
-                     * string.  Previously we didn't catch this, so people put
-                     * in all kinds of wacky interpretations of how this
-                     * could work.  We want to disallow all of them, but can't
-                     * without breaking them.  So we silently ignore strings. */
-                    *value_type = CLIENT_MINTYPE;
-                    *value_sz = 0;
-                    memset(valuebuf, 0, vbsz);
-                    return -1;
-                }
-                return -1;
+            length = *bytes;
+            bytes++;
+            if (*value_type == CLIENT_BYTEARRAY && vbsz >= length) {
+                memcpy(valuebuf, bytes, length);
+                *value_sz = length;
+                return 0;
             }
-            default:
+            return -1;
+        }
+        case CLIENT_CSTR: {
+            if (*value_type == CLIENT_BYTEARRAY && vbsz >= sym->szof) {
```

Comment:
changed the order of the ifs here, the one on the left is done a little further down on the right.